modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java

```diff
@@ -691,40 +687,103 @@ public String getKey() {
 
-    private record StatsKey(Operation operation, OperationPurpose purpose) {
+    // visible for testing
+    record StatsKey(Operation operation, OperationPurpose purpose) {
         @Override
         public String toString() {
             return purpose.getKey() + "_" + operation.getKey();
         }
     }
 
-    private static class StatsCollectors {
-        final Map<StatsKey, LongAdder> collectors = new ConcurrentHashMap<>();
+    // visible for testing
+    class StatsCollectors {
+        final Map<StatsKey, OperationMetrics> collectors = new ConcurrentHashMap<>();
 
         Map<String, Long> statsMap(boolean stateless) {
             if (stateless) {
                 return collectors.entrySet()
                     .stream()
-                    .collect(Collectors.toUnmodifiableMap(e -> e.getKey().toString(), e -> e.getValue().sum()));
+                    .collect(Collectors.toUnmodifiableMap(e -> e.getKey().toString(), e -> e.getValue().count()));
             } else {
                 Map<String, Long> normalisedStats = Arrays.stream(Operation.values()).collect(Collectors.toMap(Operation::getKey, o -> 0L));
                 collectors.forEach(
                     (key, value) -> normalisedStats.compute(
                         key.operation.getKey(),
-                        (k, current) -> Objects.requireNonNull(current) + value.sum()
+                        (k, current) -> Objects.requireNonNull(current) + value.count()
                     )
                 );
                 return Map.copyOf(normalisedStats);
             }
         }
 
-        public void onSuccessfulRequest(Operation operation, OperationPurpose purpose) {
-            collectors.computeIfAbsent(new StatsKey(operation, purpose), k -> new LongAdder()).increment();
+        public void onRequestComplete(Operation operation, OperationPurpose purpose, AzureClientProvider.RequestMetrics requestMetrics) {
+            collectors.computeIfAbsent(new StatsKey(operation, purpose), k -> new OperationMetrics(operation, purpose))
+                .onRequestComplete(requestMetrics);
         }
     }
 
+    // visible for testing
+    class OperationMetrics {
+
+        final LongAdder counter;
+        final Map<String, Object> attributes;
+
+        private OperationMetrics(Operation operation, OperationPurpose purpose) {
+            this.counter = new LongAdder();
+            this.attributes = RepositoriesMetrics.createAttributesMap(repositoryMetadata, purpose, operation.getKey());
+        }
+
+        public void onRequestComplete(AzureClientProvider.RequestMetrics requestMetrics) {
+            counter.add(requestMetrics.getRequestCount());
+
+            // range not satisfied is not retried, so we count them by checking the final response
+            if (requestMetrics.getStatusCode() == RestStatus.REQUESTED_RANGE_NOT_SATISFIED.getStatus()) {
+                repositoriesMetrics.requestRangeNotSatisfiedExceptionCounter().incrementBy(1, attributes);
```

Comment:
Does it make sense for operation to contain own subset of request metrics? And then repositoryMetrics is an aggregate of operation metrics. I think there few benefits: remove side-effect of mutating repositoriesMetrics, less contention, gives more granularity in metrics operation/request dimensions. The downside is more code to update operation metrics and repository metrics.

Reply:
I'm not sure I follow, `RepositoriesMetrics` is a singleton record that registers the APM metrics and provides access to them for places where they're recorded.
`RequestMetrics` in this place is an object containing the metrics for a single operation, `RequestMetrics` is populated over the course of executing the operation, then passed to the `RequestMetricsHandler` at the end of the request, which in this case interprets them to publish via `RepositoriesMetrics`.

I think that singleton pattern is how otel metrics are commonly used (similar to a logger) in the java SDK at least?

Reply:
The `OperationMetrics` seems to me as extra abstraction that has side-effect of changing  `RepositoriesMetrics`. I think you can remove it, something like this:

```java
class AzureRepoMetrics {
  RepositoriesMetrics repoMetrics
  Map<StatsKey, LongAdder> opsCounter
  Map<StatsKey, Attributes> opsAttributes 
  
   void addRequestMetrics(Operation operation, OperationPurpose purpose, AzureClientProvider.RequestMetrics requestMetrics) {
  ... // same as onRequestComplete
  }

  RepositoriesMetrics repoMetrics() { return repoMetrics; }
}

...
public void onRequestComplete(Operation operation, OperationPurpose purpose, AzureClientProvider.RequestMetrics requestMetrics) {
  azureRepoMetrics.addRequestMetrics(operation, purpose, requestMetrics);
}
...
```